File: it-agents/skills/code-architect/references/code-examples/java/Command.java
// Command — encapsulate a request as an object, support undo.

import java.util.ArrayDeque;
import java.util.Deque;

class Counter {
    private int value = 0;
    void add(int n) { value += n; }
    void subtract(int n) { value -= n; }
    int value() { return value; }
}

interface Action {
    void execute();
    void undo();
}

class AddCommand implements Action {
    private final Counter counter;
    private final int amount;
    AddCommand(Counter counter, int amount) { this.counter = counter; this.amount = amount; }
    public void execute() { counter.add(amount); }
    public void undo() { counter.subtract(amount); }
}

class Invoker {
    private final Deque<Action> history = new ArrayDeque<>();
    void run(Action c) { c.execute(); history.push(c); }
    void undoLast() { if (!history.isEmpty()) history.pop().undo(); }
}

public class Command {
    public static void main(String[] args) {
        Counter counter = new Counter();
        Invoker invoker = new Invoker();

        invoker.run(new AddCommand(counter, 5));
        invoker.run(new AddCommand(counter, 3));
        System.out.println("after adds: " + counter.value());

        invoker.undoLast();
        System.out.println("after undo: " + counter.value());
    }
}

File: it-agents/skills/code-architect/references/code-examples/ruby/command.rb
# frozen_string_literal: true

# Command — encapsulate a request as an object, support undo.

# Receiver: holds the state the commands act on.
class Counter
  attr_reader :value

  def initialize = @value = 0
  def add(n) = @value += n
  def sub(n) = @value -= n
end

# Command interface: #execute / #undo over a receiver.
class AddCommand
  def initialize(receiver, amount)
    @receiver = receiver
    @amount = amount
  end

  def execute = @receiver.add(@amount)
  def undo = @receiver.sub(@amount)
end

# Invoker runs commands and keeps a history for undo.
class Invoker
  def initialize = @history = []

  def run(command)
    command.execute
    @history.push(command)
  end

  def undo_last
    command = @history.pop
    command&.undo
  end
end

if __FILE__ == $PROGRAM_NAME
  counter = Counter.new
  invoker = Invoker.new

  invoker.run(AddCommand.new(counter, 5))
  invoker.run(AddCommand.new(counter, 3))
  puts "after two adds: #{counter.value}"

  invoker.undo_last
  puts "after undo: #{counter.value}"
end
